### PR TITLE
chore: Add `@HapiTest` readme, style guide, and spec cookbook

### LINKED ISSUE #14046: Document purpose of each `HederaNetwork` implementation
### Problem

We have three implementations of `HederaNetwork`:
  1. `RemoteNetwork`
  2. `SubProcessNetwork`
  3. `EmbeddedNetwork`

And `EmbeddedNetwork` has two options for its embedded node: `ConcurrentEmbeddedHedera` or `RepeatableEmbeddedHedera`.

There is some more context needed on why we have each of these configurations and where they are appropriate to use.

### Solution

Update _README.md_ to clarify this.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Closes #14046 by adding a _README.md_ and _SpecCookbook.md_ to the `test-clients` module.
 - Moves `SpecOperation` and `TargetNetworkType` to more natural packages, adds some javadoc.
 - Improves the `@BeforeAll` class-scoped overrides mechanism by reworking `SpecManager` as a `TestLifecycle` helper.

### CODE REVIEW

File: hedera-node/test-clients/README.md [hunk 1]
```diff
@@ -1,11 +1,250 @@
...
+directly against a remote network, this is uncommon. Typically, developers execute `HapiSpec`s as dynamic
+tests on the JUnit platform using JUnit Jupiter against an ephemeral four-node network whose lifetime is
+scoped to the JUnit `LauncherSession`. Such tests are marked with the `@HapiTest` meta-annotation, as correct
+execution requires registering Jupiter extensions to manage the test network lifecycle and make it the target
+of each `HapiSpec`.
+
+The default Gradle `test` task instructs these extensions to create the target network by spawning four
+child subprocesses as nodes. Meanwhile, the `testEmbedded` and `testRepeatable` tasks instruct the
+extensions to embed a single `Hedera` instance in the test process, invoking its workflows directly.
+In embedded mode, there is no hashgraph consensus because the `Platform` is replaced with a simple mock,
```
Comment (Nana-EC) [inline hedera-node/test-clients/README.md:21]: 🙌🏾

File: hedera-node/test-clients/README.md [hunk 2]
```diff
@@ -1,11 +1,250 @@
...
+1. Any qualified PR reviewer can understand the behavior being asserted **in a single reading**; and,
+2. The `@HapiTest` javadoc links any context necessary to understand the motivation for the test.
+
+### The `@HapiTest` checklist
+To help consistently meet these criteria, we use the following minimum checklist for every `@HapiTest`:
+- [x] The `@DisplayName`, when combined with the display names of any enclosing test classes, forms a single, clear
+  phrase that fully summarizes the behavior being tested.
+- [x] If the test specifies behavior from a HIP, its javadoc links to the section in the markdown file in the
+  [`hedera-improvement-proposal` repo](https://github.com/hashgraph/hedera-improvement-proposal) outlining the behavior.
+- [x] The test includes no more than seven (7) `SpecOperation`'s.
```
Comment (Nana-EC) [inline hedera-node/test-clients/README.md:181]: Q: is this a best practices recommendation or a limitation?
Also does this include setup operations or do all SpecOperation occur after setup transactions?
Some intricate flows might get close or require more operations

File: hedera-node/test-clients/README.md [hunk 3]
```diff
@@ -1,11 +1,250 @@
...
+Operations share context through their spec's `HapiSpecRegistry`. The registry maps from `String` names to
+arbitrary values; which are themselves usually references to entities on the target network. For example,
+after a `HapiCryptoCreate` operation submits a transaction that creates a new account, it stores the new
+account's id (say, `AccountID[accountNum=1001]`) in the registry with the name given by the test author
+(say, `"Alice"`). The test author can then write a following `HapiCryptoUpdate` operation that references
+the account by its stable name `"Alice"`, and not worry that the next time the spec runs the created account
+may have a different id on the target network.
+
+Besides providing the context of its registry and target network, a `HapiSpec` also supports its operations
+with some useful infrastructure components. The four most important components are,
```
Comment (Nana-EC) [inline hedera-node/test-clients/README.md:62]: Q: Is there a component for setting config override as it relates to feature flags of non `RemoteNetwork` options?

**Review by Nana-EC — COMMENTED** [review]
Great effort.
I like the direction this is taking us 😄 
Left some questions

File: hedera-node/test-clients/README.md (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/test-clients/README.md:181]: I consider it a best practice/requirement...you can see [here](https://github.com/hashgraph/hedera-services/pull/14105/files#diff-ebfd622a28cce1b2e4e1bce2078d4c8ead917f05384552f2b47c801c5a20f143R64) I just added probably a quite involved test class that goes through two software upgrades while removing and adding nodes and validating staking; but no single `@HapiTest` uses more than 6 operations.

File: hedera-node/test-clients/README.md (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/test-clients/README.md:62]: Hmm no there isn't; but that's an interesting idea. 🤔

**Review by thomas-swirlds-labs — APPROVED** [review]
LGTM @tinker-michaelj !

**Review by Neeharika-Sompalli — APPROVED** [review]
Looks great! thanks @tinker-michaelj

**Pull request merged**